### PR TITLE
Corrections to the string solver

### PR DESCRIPTION
**Pull request by allredj**
Fast-forward of #646 to `test-gen-support` branch.
No dependencies.

### CODE REVIEW

File: src/goto-programs/string_refine_preprocess.cpp [hunk 1]
```diff
@@ -692,11 +800,48 @@ void string_refine_preprocesst::make_string_function_side_effect(
...
-  const code_function_callt function_call=to_code_function_call(target->code);
-  assert(!function_call.arguments().empty());
-  string_builders[function_call.lhs()]=function_call.arguments()[0];
-  make_string_function(
-    goto_program, target, function_name, signature, true, false);
+  code_function_callt function_call=to_code_function_call(target->code);
+  source_locationt loc=function_call.source_location();
+  std::list<code_assignt> assignments;
+  exprt lhs=function_call.lhs();
+  exprt s=function_call.arguments()[0];
```
Comment (peterschrammel) [inline src/goto-programs/string_refine_preprocess.cpp]: Would be good to keep `assert(!function_call.arguments().empty());`

File: src/goto-programs/string_refine_preprocess.cpp [hunk 2]
```diff
@@ -692,11 +800,48 @@ void string_refine_preprocesst::make_string_function_side_effect(
   const irep_idt &function_name,
   const std::string &signature)
 {
-  const code_function_callt function_call=to_code_function_call(target->code);
-  assert(!function_call.arguments().empty());
-  string_builders[function_call.lhs()]=function_call.arguments()[0];
-  make_string_function(
-    goto_program, target, function_name, signature, true, false);
+  code_function_callt function_call=to_code_function_call(target->code);
```
Comment (peterschrammel) [inline src/goto-programs/string_refine_preprocess.cpp]: Make the following variables const & if you don't modify them

File: src/goto-programs/string_refine_preprocess.cpp [hunk 3]
```diff
@@ -744,16 +890,36 @@ void string_refine_preprocesst::make_to_char_array_function(
   goto_programt &goto_program, goto_programt::targett &target)
 {
   const code_function_callt &function_call=to_code_function_call(target->code);
+  source_locationt location=function_call.source_location();
 
-  assert(!function_call.arguments().empty());
+  assert(function_call.arguments().size()>=1);
```
Comment (peterschrammel) [inline src/goto-programs/string_refine_preprocess.cpp]: I'd prefer the old version

File: src/goto-programs/string_refine_preprocess.cpp [hunk 4]
```diff
@@ -974,22 +968,20 @@ exprt::operandst string_refine_preprocesst::process_arguments(
...
-      exprt arg2=make_cprover_string_assign(
-        goto_program, target, arg, location);
-      new_arguments.push_back(arg2);
+      while(arg.id()==ID_typecast)
+        arg=arg.op0();
+      if(!implements_java_char_sequence(arg.type()))
+        arg=typecast_exprt(arg, jls_ptr);
     }
+    arg=make_cprover_string_assign(goto_program, target, arg, location);
+    typet type=ns.follow(arg.type());
```
Comment (peterschrammel) [inline src/goto-programs/string_refine_preprocess.cpp]: const &

File: src/solvers/refinement/string_constraint_generator_main.cpp [hunk 5]
```diff
@@ -144,59 +181,113 @@ string_exprt string_constraint_generatort::fresh_string(
...
+      to_symbol_expr(expr),
+      to_refined_string_type(expr.type()));
+  }
+  else
+  {
+    return to_string_expr(expr);
+  }
+}
+
+string_exprt string_constraint_generatort::convert_java_string_to_string_exprt(
```
Comment (peterschrammel) [inline src/solvers/refinement/string_constraint_generator_main.cpp]: function comment block missing

File: src/solvers/refinement/string_refinement.cpp [hunk 6]
```diff
@@ -11,17 +11,38 @@ Author: Alberto Griggio, alberto.griggio@gmail.com
...
+\*******************************************************************/
 
 string_refinementt::string_refinementt(
-  const namespacet &_ns, propt &_prop, unsigned refinement_bound):
+  const namespacet &_ns,
+  propt &_prop,
+  unsigned refinement_bound):
   supert(_ns, _prop),
   use_counter_example(false),
+  do_concretizing(false),
```
Comment (peterschrammel) [inline src/solvers/refinement/string_refinement.cpp]: concrete_model is not initialised

File: src/goto-programs/string_refine_preprocess.cpp (hunk 2 above)
Comment (allredj) [inline src/goto-programs/string_refine_preprocess.cpp]: Done in `make_string_function` and `make_string_function_side_effect`.

File: src/goto-programs/string_refine_preprocess.cpp (hunk 1 above)
Comment (allredj) [inline src/goto-programs/string_refine_preprocess.cpp]: Assertion added.

File: src/goto-programs/string_refine_preprocess.cpp (hunk 4 above)
Comment (allredj) [inline src/goto-programs/string_refine_preprocess.cpp]: Corrected.

File: src/solvers/refinement/string_constraint_generator_main.cpp (hunk 5 above)
Comment (allredj) [inline src/solvers/refinement/string_constraint_generator_main.cpp]: Added.

File: src/goto-programs/string_refine_preprocess.cpp (hunk 3 above)
Comment (allredj) [inline src/goto-programs/string_refine_preprocess.cpp]: That was probably a merge error. The original version has been reinstated.

**Pull request merged**